include/openthread/platform/dnssd.h

```diff
@@ -0,0 +1,322 @@
+ * be returned in each case.
+ *
+ * @param[in] aInstance     The OpenThread instance.
+ * @param[in] aRequestId    The request ID.
+ * @param[in] aError        Error indicating the outcome of request.
+ *
+ */
+typedef void (*otPlatDnssdRegisterCallback)(otInstance *aInstance, otPlatDnssdRequestId aRequestId, otError aError);
+
+/**
+ * Represent a DNS-SD service.
+ *
+ * See `otPlatDnssdRegisterService()`, `otPlatDnssdUnregisterService()` for more details about fields in each case.
+ *
+ */
+typedef struct otPlatDnssdService
+{
+    const char        *mHostName;            ///< The host name (does not include domain name).
+    const char        *mServiceInstance;     ///< The service instance name label (not the full name).
+    const char        *mServiceName;         ///< The service name (e.g., "_mt._udp", does not include domain name).
+    const char *const *mSubTypeLabels;       ///< Array of sub-type labels (can be NULL if no label).
+    uint16_t           mSubTypeLabelsLength; ///< Length of array of sub-type labels.
+    const uint8_t     *mTxtData;             ///< Encoded TXT data bytes.
+    uint16_t           mTxtDataLength;       ///< Length of TXT data.
+    uint16_t           mPort;                ///< The service port number.
+    uint16_t           mPriority;            ///< The service priority.
+    uint16_t           mWeight;              ///< The service weight.
+    uint32_t           mTtl;                 ///< The service TTL in seconds.
+} otPlatDnssdService;
+
+/**
+ * Represent a DNS-SD host.
+ *
+ * See `otPlatDnssdRegisterHost()`, `otPlatDnssdUnregisterHost()` for more details about fields in each case.
+ *
+ */
+typedef struct otPlatDnssdHost
+{
+    const char         *mHostName;     ///< The host name (does not include domain name).
+    const otIp6Address *mAddresses;    ///< Array of IPv6 host addresses.
+    uint16_t            mNumAddresses; ///< Number of entries in @p mAddresses array.
+    uint32_t            mTtl;          ///< The host TTL in seconds.
+} otPlatDnssdHost;
+
+/**
+ * Callback from platform to notify OpenThread of state changes to DNS-SD module.
+ *
+ * OpenThread stack will call `otPlatDnssdGetState()` (from this callback or later) to get the new state. The platform
+ * MUST therefore ensure that the returned state from `otPlatDnssdGetState()` is updated before calling this.
+ *
+ * @param[in] aInstance The OpenThread instance structure.
+ *
+ */
+extern void otPlatDnssdStateHandleStateChange(otInstance *aInstance);
+
+/**
+ * Gets the current state of DNS-SD module.
+ *
+ * Platform MUST notify OpenThread whenever its state gets changed by invoking `otPlatDnssdStateHandleStateChange()`.
+ *
+ * @param[in] aInstance The OpenThread instance.
+ *
+ * @returns The current state of DNS-SD module.
+ *
+ */
+otPlatDnssdState otPlatDnssdGetState(otInstance *aInstance);
+
+/**
+ * Registers or updates a service on the infrastructure network's DNS-SD module.
```

Comment:
Should we pass in the InfraIf index to make it clear that this service MUST be published on the infrastructure network?

Another reason is that we want to support changing infra if at runtime on Android, and it would be simpler to support this case if the Infra If index is provided here.

Reply:
I think it may be better to not include the infra-if index in the DNSSD platform APIs.

If we include the infra-if index in the `otPlatDnssd` APIs, we need the platform to somehow tell us which infra-if index to use. Within the OT core itself, we don't really need to know the infra-if index (other than passing it back as a parameter in platform APIs).

It may be simpler to assume that the platform should know the infra-netif (or maybe even a set of netifs) that it should do mDNS/DNS-SD on.

To support the use case where the infra-if changes at runtime with the DNS-SD platform APIs, we can signal such a change from the platform as a "state change" using `otPlatDnssdStateHandleStateChange()`. This should then trigger the OT stack to re-register all mDNS/DNS-SD entries (services/hosts) on new infra-if. 

One more related thought on this:
- In the `otPlatInfraIf` APIs, we do have the infra-If index as input, but this is tied to the BR and specifically `RoutingManager` functionality.
- The way we learn the infra-if index is actually through a public BR API `otBorderRoutingInit()`
- If we try to use the same infra-if index for DNSSD platform API, we indirectly restrict the use of the `otPlatDnssd` APIs to the use of the `RoutingManager` (which may not be good).

Thoughts?

Reply:
> - In the otPlatInfraIf APIs, we do have the infra-If index as input, but this is tied to the BR and specifically RoutingManager functionality.
> - The way we learn the infra-if index is actually through a public BR API otBorderRoutingInit()
> - If we try to use the same infra-if index for DNSSD platform API, we indirectly restrict the use of the otPlatDnssd APIs to the use of the RoutingManager (which may not be good).

The Infra interface passed to otPlatDnssd APIs doesn't necessary to be always the AIL selected for border routing. The otPlatDnssd APIs can be reused for TREL which can be functioning on another infra interface. For example, RoutingManager works on wlan0 but TREL works on eth0. This is allowed, right?

For border routing, it's better to advertise SRP service on an AIL only when we can provide connectivity between Thread and this AIL. If we don't pass a Infra interface parameter with the otPlatDnssd APIs, the implementations may just advertise the services on all AILs available on this host (OTBR today is doing so) and we certainly want to help avoid this by telling the implementer which exact infra interface he/she should use. Thoughts?

Reply:
Thinking more on this, I agree it may be useful to sepcify the infra if index. 
Will add it as `mInfraIfIndex` variable in the `otPlatDnssdHost`, `Servcie` or `Key` structs (instead of an API input).

Reply:
Thanks @abtink !

> Will add it as mInfraIfIndex variable in the otPlatDnssdHost, Servcie or Key structs (instead of an API input).

This works for me, though I don't think `mInfraIfIndex` look like a property of a Host or Service - for example, if we change the advertising network interface, we shouldn't need to make update to the Host or Service objects.

Reply:
My plan is to reuse the `otPlatDnssdService` and `otPlatDnssdHost` definitions in the "subscribe/browse" DNS-SD APIs (will add later). for use by discovery proxy, etc. The `mInfraIfIndex` property of these structs can be helpful for these APIs as well, and it can be used to learn which interface the entry was discovered on. We have something similar in the `otbr::Mdns::Publisher::DiscoveredInstanceInfo` struct.

Also, these types are used as temporary stack objects that are constructed to pass to the API. We do not retain them afterwards.

Reply:
Added this in the new pushed commit.